sopel/plugins/rules.py

```diff
@@ -1158,36 +1166,63 @@ def is_global_rate_limited(self):
         rate_limit = datetime.timedelta(seconds=self._global_rate_limit)
         return self._metrics_global.is_limited(now - rate_limit)
 
+    def rate_limit_info(self, trigger: Trigger) -> RateLimitCheckInfo:
+        if trigger.admin or self.is_unblockable():
+            return RateLimitCheckInfo()
+
+        # TODO: these functions call now() and probably shouldn't
+        if self.is_user_rate_limited(trigger.nick):
+            template = self.get_user_rate_message(trigger.nick)
+        elif self.is_channel_rate_limited(trigger.sender):
+            template = self.get_channel_rate_message(trigger.nick, trigger.sender)
+        elif self.is_global_rate_limited():
+            template = self.get_global_rate_message(trigger.nick)
```

Comment:
If the getters are not public API that we want to preserve, their logic can be moved here

```suggestion
        if self.is_user_rate_limited(trigger.nick):
            template = self._user_rate_message or self._default_rate_message
        elif self.is_channel_rate_limited(trigger.sender):
            template = self._channel_rate_message or self._default_rate_message
        elif self.is_global_rate_limited():
            template = self._global_rate_message or self._default_rate_message
```

Reply:
Similarly, the business of the `is_*_rate_limited()` methods might find a home in this function as well, since they are only being called internally to `Rule` now

Reply:
None of `Rule`'s interface is really "public API" in the sense that we encourage or expect plugin developers to use it. One look at the `.. important::` at the top of the docs page that covers `Rule` & friends will tell you that. ("Do **not** build your plugin based on what is here, you do **not** need to.")

No need to plan around deprecation cycles for this stuff. No one outside of Sopel core is supposed to be using it, and if they are we can't be held responsible for their stuff breaking. _They_ made the choice to use clearly marked internal machinery. 🤷‍♂️

Reply:
For the sake of posterity, the `get_*` family are removed in 3a7063b, but the `is_*` family are staying around as possible future public API. I'm sure this PR will make some kind of changes to them, but they aren't going anywhere, so this thread is now resolved.